### PR TITLE
[FIRRTL] Store port annotations as result attributes for InstanceOp

### LINKED ISSUE #1228: [FIRRTL] Use Result Attributes to Store Instance/Memory Port Annotations
Currently, annotations associated with an instance port (or a module port) wind up stored as attributes on the declaration. This should be changed to be result attributes since ports are being represented as return values.

This aligns the representation of module/instance ports with how we represent module port annotations as argument attributes.

```scala
circuit Foo: %[[
  {"fooPort":null,"target":"~Foo|Foo>a"},
  {"barPort":null,"target":"~Foo|Foo>bar.a"}
]]
  module Bar:
    input a: UInt<1>
  module Foo:
    input a: UInt<1>

    inst bar of Bar
    bar.a <= a
```

This is currently parsed into different representations for the annotation on `Foo>a` compared to `Foo>bar.a`:
```mlir
module  {
  firrtl.circuit "Foo"  {
    firrtl.module @Bar(in %a: !firrtl.uint<1>) {
    }
    firrtl.module @Foo(in %a: !firrtl.uint<1> {firrtl.annotations = [{fooPort}]}) {
      %bar_a = firrtl.instance @Bar  {annotations = [{barPort, target = [".a"]}], name = "bar"} : !firrtl.flip<uint<1>>
      firrtl.connect %bar_a, %a : !firrtl.flip<uint<1>>, !firrtl.uint<1>
    }
  }
}
```

This should be changed to:
```mlir
%bar_a = firrtl.instance @Bar  {name = "bar"} : !firrtl.flip<uint<1>> {firrtl.annotations = [{barPort}]}
```

### PR DESCRIPTION
**Pull request by hanchenye**
Resolved #1228

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp
Comment (lattner) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:952]: The consequence of this is that `getPortAnnotation` will never return null.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp
Comment (lattner) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:967]: Please don't add this helper method.  The "getReferencedModule()" call is O(n) in the size of the CircuitOp.  We should do everything in terms of port indices, to avoid algorithmic cliffs.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp
Comment (lattner) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:1033]: Plz make this attribute required to eliminate the if.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp
Comment (lattner) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:2472]: Please set the port annotations once at the end instead of repeatedly calling setPortAnnotation (and also, please work by index, not by name).  The difference here is that each intermediate "set" will create a new ArrayAttr on the outside level and MLIR uniques these, so they will never get deallocated.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp
Comment (hanchenye) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:956]: I think we can do either way. The customized printer and parser is for pretty asm format: printing each port annotation after the corresponding port type.

Comment (lattner) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:956]: I'd recommend keeping it simple and staying with the default format unless port annotations are common.  If they're common, then having a pretty format makes a lot of sense!

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp
Comment (lattner) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:872]: This can be simplified a bit to:

`SmallVector<Attribute, 16> portAnnotationsVec(resultTypes.size(), builder.getArrayAttr({}));`

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp
Comment (lattner) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:2474]: lookup can return null here, which isn't valid to put into an ArrayAttr.  If you know that all names are present, then please add an `assert(annotation && "rationale")` here, otherwise please handle the null

**Review by lattner — APPROVED** [review]
Looks great, thanks again!

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp
Comment (hanchenye) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:956]: Sure, I think currently the only other case is `MemOp`. I'll just use the default format for these two.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp
Comment (hanchenye) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:2474]: We are good here as the `lookup` will construct an empty `SmallVector`, which is valid to construct the `ArrayAttr`.

Comment (lattner) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:2474]: oh nice, thanks!

**Review by seldridge — COMMENTED** [review]
Correct me if I'm wrong, but this is storing the port annotations in a custom `portAnnotations` attribute and not as result attributes? I had envisioned this as actually using the result attributes (which is similar to how it works for ports where annotations are actual argument attributes).

The benefit of doing it as result attributes is that the printing/parsing isn't going to be so complicated. The possible downside is that the annotations are then stored sparsely. However, I don't anticipate having extremely dense annotations on instance ports.

What do you think? (Or did I miss something? 😄)

What do you think?

**Comment by hanchenye** [comment]
@seldridge The result attributes related hooks belong to the `FunctionLikeTrait`, thus cannot be used here. It may worthwhile to introduce a `HasResultAttr` trait in the upstream MLIR for tackling this. But you are right, the data structure under the hood of `FunctionLikeTrait` is also a `ArrayAttr` like what we do here.

**Comment by hanchenye** [comment]
And yeah, we may also want to provide some methods for the printing and parsing.

**Comment by seldridge** [comment]
I don't think this has to mix in `FunctionLikeTrait`. The definition of stuff like mlir::function_like_impl::*` should take an arbitrary operation and just do the lookup using the correct name for result attrs (`"res_attrs"`). It's kind of janky, but is an approach that could work here. I don't know how that interacts with the default printer/parser.

**Comment by hanchenye** [comment]
Auh, we should be able to reuse some methods in `function_like_impl`, which will help on the `set/getPortAnnotation` related methods. I suspect there're only methods like `printFunctionSignature` to help the printing and parsing of function-like ops, no interacts with the default printer and parser.

**Comment by hanchenye** [comment]
@seldridge Didn't find useful printer/parser hooks :facepalm: Use default printer and parser now.

I'll merge this if nothing needs to be changed.

**Comment by seldridge** [comment]
I'd like to get Lattner to weight in on the storage here before this lands.

@lattner: Do you have an opinion on the dense storage here `ArrayAttr` keyed on the custom `portAnnotations` vs. sparse storage keyed using the special result attrs string? I had envisioned the latter as the implementation as it follows how module ports work (and isn't creating a mostly empty `ArrayAttr` where I expect port annotations are somewhat rare). However, the former works and I'm fine with it if you are.

**Comment by hanchenye** [comment]
Actually if I understand this correctly, the `FModuleOp` port annotations are also stored densely. The `res_attrs` will store an empty `ArrayAttr` for the ports without annotations.

**Comment by seldridge** [comment]
Haha, you're right! Good to go from me.

Example:
```scala
circuit Foo: %[[{"a":null,"target":"~Foo|Foo>b"}]]
  module Foo:
    input a: UInt<1>
    input b: UInt<1>
```

Produces:
```mlir
"module"() ( {
  "firrtl.circuit"() ( {
    "firrtl.module"() ( {
    ^bb0(%a: !firrtl.uint<1>, %b: !firrtl.uint<1>):  // no predecessors
    }) {annotations = [], arg_attrs = [{}, {firrtl.annotations = [{a}]}], portDirections = 0 : i2, portNames = ["a", "b"], sym_name = "Foo", type = (!firrtl.uint<1>, !firrtl.uint<1>) -> ()} : () -> ()
  }) {annotations = [], name = "Foo"} : () -> ()
}) : () -> ()
```

So, the solution in this PR is to flatten one of these levels and is more performant than using the result attrs. This should be fine and may motivate switching to similar storage for module ports to avoid the `arg_attrs` redirection.

**Comment by hanchenye** [comment]
Auh right, this PR doesn't have the `ArrayAttr` to `DictionaryAttr` conversion, which is required by the `function_like_impl` stuffs.

**Pull request merged**
